spec/fastsync/version2/scheduler.tla

```diff
@@ -0,0 +1,580 @@
+    [err |-> noErr, val |-> newSc]
+
+markReceived(sc, peerID, h) ==
+  IF peerID \notin sc.peers \/ sc.peerStates[peerID] /= "peerStateReady" THEN
+    [err |-> "errBadPeerState", val |-> sc]
+  ELSE IF getStateAtHeight(sc, h) /= "blockStatePending" \/ sc.pendingBlocks[h] /= peerID THEN
+    [err |-> "errBadPeer", val |-> sc]
+  ELSE
+    LET newSc == [sc EXCEPT
+      !.blockStates = [sc.blockStates EXCEPT ![h] = "blockStateReceived"],
+      !.pendingBlocks = [sc.pendingBlocks EXCEPT ![h] = None],
+      !.receivedBlocks = [sc.receivedBlocks EXCEPT ![h] = peerID]] IN
+    [err |-> noErr, val |-> newSc]
+
+markProcessed(sc, h) ==
+  IF getStateAtHeight(sc, h) /= "blockStateReceived" THEN
+    [err |-> "errProcessedBlockEv", val |-> sc]
+  ELSE
+    LET newSc == [sc EXCEPT
+      !.blockStates = [sc.blockStates EXCEPT ![h] = "blockStateProcessed"],
+      !.receivedBlocks = [sc.receivedBlocks EXCEPT ![h] = None],
+      !.height = sc.height + 1] IN
+    [err |-> noErr, val |-> newSc]
+
+allBlocksProcessed(sc) ==
+  IF sc.peers = {} THEN
+    FALSE
+  ELSE
+    LET maxH == maxHeightScheduler(sc) IN
+    maxH > 0 /\ (sc.height >= maxH)
+ 
+highPeers(sc, minH) == {p \in sc.peers: sc.peerHeights[p] >= minH}
+
+(* ----------------------------------------------------------------------------------------------*)
+(* The behavior of the scheduler state machine                                                   *)
+(* See scheduler.go                                                                              *)
+(* https://github.com/tendermint/tendermint/tree/brapse/blockchain-v2-riri-reactor-2/scheduler.go*)
+(* ----------------------------------------------------------------------------------------------*)
+InitSc ==
+  /\ scRunning = TRUE
+  /\ outEvent = noEvent
+  /\ \E startHeight \in Heights:
+    scheduler = [
+      initHeight |-> startHeight,
+      height |-> startHeight,
+      peers |-> {},
+      peerHeights |-> [p \in PeerIDs |-> None],
+      peerStates |-> [p \in PeerIDs |-> "peerStateUnknown"],
+      blocks |-> {},
+      blockStates |-> [h \in Heights |-> "blockStateUnknown"],
+      pendingBlocks |-> [h \in Heights |-> None],
+      receivedBlocks |-> [h \in Heights |-> None]
+      ]
+     
+handleAddNewPeer ==
+  /\ inEvent.type = "bcAddNewPeer"
+  /\ LET res == addPeer(scheduler, inEvent.peerID) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+       /\ UNCHANGED <<scheduler>>
+     ELSE
+       /\ scheduler' = res.val
+       /\ UNCHANGED outEvent
+  /\ UNCHANGED scRunning
+       
+handleRemovePeer ==
+  /\ inEvent.type = "bcRemovePeer"
+  /\ LET res == removePeer(scheduler, inEvent.peerID) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+       /\ UNCHANGED scheduler
+     ELSE
+       /\ scheduler' = res.val
+       /\ IF allBlocksProcessed(scheduler') THEN
```

Comment:
Maybe you can pass res.val as a parameter here and use variables with prime only on the left side of relations :)